File: XIRHXQ_Robot.m
classdef XIRHXQ_Robot < handle
    properties
        n(1, 1) {mustBeInteger}
        v XIRHXQ_Point
        p XIRHXQ_Point
        s(1, 1) {mustBePositive} = 1
    end
    methods
        function obj = XIRHXQ_Robot(varargin)
            if nargin > 3 || nargin < 2
                error('Initialize Error: Invalid Input');
            else
                obj.n = varargin{1};
                if isnumeric(varargin{2})
                    for i = obj.n : -1 : 1
                        obj.p(i) = XIRHXQ_Point(varargin{2}(i, :));
                    end
                else
                    obj.p = varargin{2};
                end
                if nargin == 3
                    if isnumeric(varargin{3})
                        for i = obj.n : -1 : 1
                            obj.v(i) = XIRHXQ_Point(varargin{3}(i, :));
                        end
                    else
                        obj.v = varargin{3};
                    end
                end
            end
        end
        function obj = set_saturation(obj, saturation)
            obj.s = saturation;
        end
        function obj = set_velocity(obj, velo)
            if isnumeric(velo)
                if size(velo, 1) ~= obj.n || size(velo, 2) ~= 2
                    error('Wrong size of velocity input');
                end
                for i = 1:obj.n
                    obj.v(i) = XIRHXQ_Point(velo(i, :));
                end
            else
                obj.v = velo;
            end
        end
        function obj = time_forward(obj, time)
            for i = 1:obj.n
                obj.p(i) = obj.p(i) + time * obj.v(i).saturation(obj.s);
            end
        end
    end
end

File: cpp_file/XIRHXQ_Robot.cpp
// Created by XIRHXQ on 12/30/21.

#include "utils.h"
#include "XIRHXQ_Robot.h"

XIRHXQ_Robot::XIRHXQ_Robot() {
    p = XIRHXQ_Point();
    v = XIRHXQ_Point();
}

XIRHXQ_Robot::XIRHXQ_Robot(double _x, double _y) {
    p = XIRHXQ_Point(_x, _y);
    v = XIRHXQ_Point();
}

XIRHXQ_Robot::XIRHXQ_Robot(XIRHXQ_Point &_p) {
    p = _p;
    v = XIRHXQ_Point();
}

XIRHXQ_Robot::XIRHXQ_Robot(XIRHXQ_Point &_p, XIRHXQ_Point &_v) {
    p = _p;
    v = _v;
}

XIRHXQ_Robot::XIRHXQ_Robot(XIRHXQ_Point &_p, XIRHXQ_Point &_v, double _s) {
    p = _p;
    v = _v;
    s = _s;
}

XIRHXQ_Robot::XIRHXQ_Robot(XIRHXQ_Point &_p, double _s) {
    p = _p;
    v = XIRHXQ_Point();
    s = _s;
}

void XIRHXQ_Robot::output() {
    printf("An XIRHXQ_Robot @ (%.4lf, %.4lf) with velocity (%.4lf, %.4lf)\n", p.x, p.y, v.x, v.y);
}


void XIRHXQ_Robot::set_position(const XIRHXQ_Point &_p) {
    p = _p;
}

void XIRHXQ_Robot::set_velocity(const XIRHXQ_Point &_v) {
    v = _v;
}

void XIRHXQ_Robot::set_velocity(double _x, double _y) {
    v = XIRHXQ_Point(_x, _y);
}

void XIRHXQ_Robot::set_saturation(double _s) {
    s = _s;
}

void XIRHXQ_Robot::time_forward(double _t) {
    p = p + v.saturation(s) * _t;
}

const XIRHXQ_Point &XIRHXQ_Robot::getP() const {
    return p;
}